### ISSUE TITLE
[FR] Add ordered categorical distribution

### ISSUE THREAD

**Issue by ecotner** [pyro-ppl/pyro#2571]
### Issue Description
I'm going through the Statistical Rethinking textbook by McElreath and in chapter 12 the author discusses how to handle ordered categorical variables in a sensible way. I tried to look through the lists of distributions available in both `torch` and `pyro`, but could not find anything similar to the `dordlogit` (aka ordered-logit) distribution that he describes in the text. I realize that it's just a re-parametrization of the categorical distribution, but I think it might be a useful utility to include.

I tried to recreate the distribution myself; does this seem like a sensible implementation, and would it be something you would want to include in `pyro`?

### Code Snippet
I'm no expert in implementing torch/pyro distributions, but this seems to do the job, and I have used it successfully to do the trolley problem example in the textbook using NUTS. There are probably other methods/attributes that are missing for a full implementation.

```python
import torch
from torch.distributions import constraints
from pyro.distributions import TorchDistribution, Categorical

class OrderedCategorical(TorchDistribution):
    """
    Alternative parametrization of the distribution over a categorical variable.
    
    Instead of the typical parametrization of a categorical variable in terms
    of the probability mass of the individual categories ``p``, this provides an
    alternative that is useful in specifying ordered categorical models. This
    accepts a list of ``cutpoints`` which are a (potentially initially unordered)
    vector of real numbers denoting baseline cumulative log-odds of the individual
    categories, and a model vector ``phi`` which modifies the baselines for each
    sample individually.
    
    These cumulative log-odds are then transformed into a discrete cumulative
    probability distribution, that is finally differenced to return the probability
    mass function ``p`` that specifies the categorical distribution.
    """
    support = constraints.nonnegative_integer
    arg_constraints = {"phi": constraints.real, "cutpoints": constraints.real}
    has_rsample = False
    
    def __init__(self, phi, cutpoints):
        assert len(cutpoints.shape) == 1 # cutpoints must be 1d vector
        assert len(phi.shape) == 1 # model terms must be 1d vector of samples
        N, K = phi.shape[0], cutpoints.shape[0]+1
        cutpoints = torch.sort(cutpoints).values.reshape(1, -1)  # sort and reshape for broadcasting
        q = torch.sigmoid(cutpoints - phi.reshape(-1, 1))  # cumulative probabilities
        # turn cumulative probabilities into probability mass of categories
        p = torch.zeros((N, K)) # (batch/sample dim, categories)
        p[:,0] = q[:,0]
        p[:,1:-1] = (q - torch.roll(q, 1, dims=1))[:,1:]
        p[:,-1] = 1 - q[:,-1]
#         self.cum_prob = q
        self.dist = Categorical(p)
        
    def sample(self, *args, **kwargs):
        return self.dist.sample(*args, **kwargs)
    
    def log_prob(self, *args, **kwargs):
        return self.dist.log_prob(*args, **kwargs)
    
    @property
    def _event_shape(self):
        return self.dist._event_shape
```

**Comment by fehiepsi** [comment]
@ecotner I like to have this distribution in Pyro. We also have [OrderedLogistic](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/discrete.py#L400) implemented in NumPyro, which is quite similar to your implementation but has ordered constraint for `cutpoints` (probably this is easier to interpret?) instead of sorting it in the constructor. It might be better to avoid identification problem here (in your implementation, a permutation of `cutpoints` gives the same distribution) and to have the same interface across Pyro and NumPyro. WDYT?

**Comment by ecotner** [comment]
Yes I see what you mean; forcing the input to be already sorted reduces confusion about what the cutpoints mean and which elements are related to which categories.

Maybe as a compromise we could add a kwarg `sort_cutpoints=False` that if left as default, will throw an error when the cutpoints are out of order, and if changed to `True`, will take care of the sorting for you (for lazy people like myself :smile:)?

Perhaps this is an idea for a separate FR/PR, but one could also use this in conjunction with a new `SortTransform` that would order the cutpoints for you like

```python
def model(*args):
    ...
    cutpoints = pyro.sample("cutpoints", TransformedDistribution(Normal(0., 1.).expand([K-1]), SortTransform()))
    dist = OrderedCategorical(phi, cutpoints, sort_cutpoints=False)
    ...
```
probably very similar to the `Permute` transform that already exists.

**Comment by ecotner** [comment]
Also, I noticed the naming differences `phi` -> `predictors` and `OrderedCategorical` -> `OrderedLogit` between my example and your example in `numpyro`; shall we use the latter naming scheme for better synchronicity? I have no preference.

**Comment by fehiepsi** [comment]
Haha, I think we have similar point of view here (except for using `sort` which I prefer to avoid because identification will be a big problem for mixing MCMC chains - e.g. chain 1 concentrates around cutpoints `(0, 1)`, chain 2 concentrates around cutpoints `(1, 0)`). In NumPyro, there is `OrderedTransform` to be used for TransformedDistribution as above (I mainly use this) and `constraints.ordered_vector` to be used to define improper distribution over `cutpoints` (there are some examples in [this PR](https://github.com/pyro-ppl/numpyro/pull/619)).

**Comment by fritzo** [comment]
:+1: to matching the NumPyro & Pyro interfaces.

Just to provide another datapoint: I found when designing the `CoalescentTimes` interface that it seemed most natural for the times to be ordered, but I also implemented an unordered `CoalescentTimesLikelihood` (which is kind of a funsor transpose, `p(times|-)` rather than `p(-|rate)`) where we sort internally. The constraint I implemented there takes an `ordered` kwarg, and can be used either way. https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/coalescent.py

**Comment by ecotner** [comment]
@fehiepsi I've been getting deeper into the exercises in Statistical Rethinking, and I have run into the non-identifiability problem you mentioned; it is much more troublesome than I would have thought... basically wrecking my model, even when using a single chain! So I have to say I agree with you 100% now, sorting is definitely not the way to go, either inside nor outside the distribution's `__init__`.

What I have been doing to get around this is to transform the cutpoint vector into a cumulative sum of positive variables to ensure their monotonically-increasing order like
```python
cutpoints = pyro.sample("cutpoints", Normal(-1.5, 1.5).expand([K-1]).to_event(0))
cutpoints = cutpoints.exp().cumsum(dim=-1).log()
```
(the `log` is there to try and undo any explosive increase due to the `exp`; I assume this can be done in a more numerically stable way, but it gets the job done for now). I noticed that this is similar, but not exactly the same transformation that is used in the `OrderedTransform` in numpyro, which itself is suggested by the Stan manual. Do you know if there is any theoretical support for using these types of tranformations, or is it just one of convenience/simplicity?

Given the above, I now think that allowing the user to pass in unordered cutpoints to an `OrderedLogit` distribution and have some ordering transformation applied automatically might be confusing, since the user will have to go back through the documentation and try to understand what transformation is being applied if something goes wrong. So I am now leaning towards enforcing ordering on the input itself like you were suggesting, and having the user be more cognizant of the transformations they need to do to make an ordered vector of cutpoints.

@fritzo am I to interpret your comment as being in favor of auto-ordering cutpoints (in a way that doesn't hurt identifiability) as an option? What do you think given my thoughts in the previous paragraph?

**Comment by fehiepsi** [comment]
> there is any theoretical support for using these types of tranformations

I think it is easier to interpret. By giving positive-support priors on `foo` and define `cutpoints = c = foo.cumsum()`, we can interpret that those are priors for the lags of cutpoints (or levels) (c_[i + 1] - c_[i]). With your transform, I interpret it as you want to set "LogNormal" prior for the lags of the exponentials of cutpoints.

**Comment by fritzo** [comment]
> I am now leaning towards enforcing ordering on the input

Sounds good.

> @fritzo am I to interpret your comment as being in favor of auto-ordering cutpoints ...?

I don't have a preference; I merely wanted to provide some additional examples. Actually in the examples to which I pointed, the cutpoints were a parameter rather than a distribution value (i.e. random variable), thus we didn't have an issue with identifiability.

**Comment by ecotner** [comment]
Hey, sorry I went MIA for so long, work/life kind of got in the way for a while haha. I made a PR (#2651) for this if you guys would like to check it out. I got rid of the sorting operation so there is no identifiability issue that @fehiepsi was concerned about.

**Comment by ecotner** [comment]
Thanks for all the advice implementing the `OrderedLogistic` distribution @fehiepsi and @fritzo, I learned a lot from that. Was also thinking we might want to complement it with an `OrderedTransform` for cutpoints [like in numpyro](http://num.pyro.ai/en/stable/_modules/numpyro/distributions/transforms.html#OrderedTransform). Does that sound like a good idea?

**Comment by fehiepsi** [comment]
@ecotner Yeah, having OrderedTransform would be great. It will allow us to do inference over the cutpoints parameter.

**Comment by fehiepsi** [comment]
This is addressed in #2662. Thanks, @ecotner!

**Comment by jonmumm** [comment]
Wanted to say I am using this in a personal project and it's very helpful—thank you all (also going through Statistical Relearning)
